File: OlderCodes/Ccodes/Daub4Compress_2.c
/* 
************************************************************************
*  Daub4Compress.c
  1D Daubechies 4-coefficient wavelet transform based on Press et al   *
*                                                                      *
*  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; & CC BORDEIANU, Univ Bucharest, 2008
   Support by National Science Foundation                              
*                                                                      *
************************************************************************
*/

#include<stdio.h>
#include<math.h>
#include<stdlib.h>
 main() {
// initial data in initial.dat, compressed image in daubc.dat
 
FILE *output1,*output2;
      output1=fopen("initial.dat","w");
      output2=fopen("daubc.dat","w");

int i;
int n=1024;
void wt1(double x[], int n, int sign);
void daub4(double x[], int n, int sign);
double x[1025];//data vector
//initial data via  arbitrary function
for (i=1;i<=n;i++)
    { x[i] = (1.0/1204)* (i*i-3.0*i); 
      fprintf(output1,"%d\t%f\n",i,x[i]); }
wt1(x,n,1);   //DWT
wt1(x,n,-1);   //inverse DWT
for (i=1; i <= n; i++) fprintf(output2,"%d\t%f\n",i,x[i]);
}
void wt1(double x[], int n, int sign)
//1D DWT via pyramid algorithm, replace x[] by TF
//sign = +- 1 for TF/inverse, n = 2^N, nn>>1 = nn/2
{
void daub4(double x[], int n, int sign);
int nn;
if (n < 4) return;
if (sign >= 0) {for (nn=n;nn>=4;nn>>=1) daub4(x,nn,sign);}
else { for (nn=4;nn<=n;nn<<=1) daub4(x,nn,sign); }
}
void daub4(double x[], int n, int sign)
//Apply 4-coefficient matrix, or transpose for sign=-1
{       //Daubechies 4 coefficients
double C0 =0.4829629131445341;  
double C1 =0.8365163037378079;
double C2 =0.2241438680420134;
double C3 =-0.1294095225512604;
double temp[1025];      
int nh,nh1,i,j;
double tol;             // threshold
tol=(40.0/100.0)*868.0;
if (n < 4) return;
nh1=(nh=n >> 1)+1;  // L and H filters
if (sign >= 0) {for (i=1,j=1;j<=n-3;j+=2,i++) 
{ temp[i]= C0*x[j]+C1*x[j+1]+C2*x[j+2]+C3*x[j+3]; 
  temp[i+nh] = C3*x[j]-C2*x[j+1]+C1*x[j+2]-C0*x[j+3]; 
  if (abs(temp[i+nh]) < tol)  temp[i+nh] = 0.0;  }
temp[i] = C0*x[n-1]+C1*x[n]+C2*x[1]+C3*x[2]; 
temp[i+nh] = C3*x[n-1]-C2*x[n]+C1*x[1]-C0*x[2]; 
if (abs(temp[i+nh]) < tol) temp[i+nh] = 0.0;
} 
else //inverse DWT
{ temp[1] = C2*x[nh]+C1*x[n]+C0*x[1]+C3*x[nh1]; 
  temp[2] = C3*x[nh]-C0*x[n]+C1*x[1]-C2*x[nh1]; 
  for (i=1,j=3;i<nh;i++) {
temp[j++] = C2*x[i]+C1*x[i+nh]+C0*x[i+1]+C3*x[i+nh1]; 
temp[j++] = C3*x[i]-C0*x[i+nh]+C1*x[i+1]-C2*x[i+nh1]; }
}
for (i=1;i<=n;i++) { 
         x[i]=temp[i];
}
}  //main

File: OlderCodes/JavaCodes/Daub4Compress2.java
/* From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
   Support by National Science Foundation                              
   */
// Daub4Compress2.java: Wavelet compression with Daub4 wavelets
import java.io.*; 
import java.util.*; 
  
public class Daub4Compress2  {
  static double c0, c1, c2, c3;                             // Global Daub4 variables
  static int n;                                                    // 2^n data points
  
  public static void pyram(double f[], int n, int sign) {
 // 1D DWT via pyramid algorithm replaces x[] by TF, (sign = +/-1 for TF/inverse
    int nd, nend; 
    double sq3, fsq2;                                            // sqrt(3), 4sqrt(2)
    if (n  <  4) return;                                              // Too few data
    sq3 = Math.sqrt(3);     
    fsq2 = 4.*Math.sqrt(2); 
    c0 = (1. + sq3)/fsq2;                                 // Daubechies 4 coefficents
    c1 = (3. + sq3)/fsq2; 
    c2 = (3.-sq3)/fsq2; 
    c3 = (1.-sq3)/fsq2; 
    nend = 4;                                 // Controls output (1:1024, 2:512, etc)
    if (sign >= 0) {for ( nd=n; nd>=nend; nd/=2) daube4(f, nd, sign);}
    else {for ( nd=4; nd <= n; nd*=2) daube4(f, nd, sign); }
  }                                                                          // Pyram
  
  public static void daube4(double f[], int n, int sign){
    double tr[] = new double [n + 1];                                // Temp variable
    int i, j; 
    int mp, mp1; 
    double tol = (40./100.)*868. ;                             // Arbitrary threshold
     // Donoho's universal thresholding for earthquake data 
     // tol = var*sqrt(2*log(n)), var = estimate of noise variance
     // deselect the following line when process earthquake data
     // double tol = 1.1*0.013481721*Math.sqrt(2*Math.log(n));    
    if (n  <  4) return; 
    mp = n/2;                                                    // Midpoint of array
    mp1 = mp + 1;                                                // Midpoint plus one
    if (sign >= 0) {            // DWT
      j = 1; 
      for ( i=1; j <= n-3; i++ ) {
        tr[i] = c0*f[j] + c1*f[j + 1] + c2*f[j + 2] + c3*f[j + 3]; 
        tr[i + mp] = c3*f[j]-c2*f[j + 1] + c1*f[j + 2]-c0*f[j + 3];  
        if (Math.abs(tr[i + mp])  <  tol)  tr[i + mp] = 0. ;               // d coeff
        j += 2; 
      }
      tr[i] = c0*f[n-1] + c1*f[n] + c2*f[1] + c3*f[2];                    // Low-pass
      tr[i + mp] = c3*f[n-1]-c2*f[n] + c1*f[1]-c0*f[2];                  // High-pass
      if (Math.abs(tr[i + mp])  <  tol) tr[i + mp] = 0. ; 
      }  
			else {                                                          // Inverse DWT
      tr[2] = c3*f[mp]-c0*f[n] + c1*f[1]-c2*f[mp1];               // High-pass filter
      for ( i=1, j=3; i < mp; i++ )  {
        tr[j] = c2*f[i] + c1*f[i + mp] + c0*f[i + 1] + c3*f[i + mp1];
        j += 1;                                          // Upsampling c coefficients
        tr[j] = c3*f[i]-c0*f[i + mp] + c1*f[i + 1]-c2*f[i + mp1];
        j += 1;                                          // Upsampling d coefficients
      }                                                                        // For
    }                                                                         // Else
    for ( i=1; i <= n; i++ )   f[i] = tr[i];                        // Copy transform
  }

  public static void main(String[] argv) throws IOException, FileNotFoundException {
     // initial data in indata.dat, compressed image in outdata
    PrintWriter w = new PrintWriter (new FileOutputStream("indata.dat"), true); 
    PrintWriter q = new PrintWriter (new FileOutputStream("outdata.dat"), true);   
    int n = 1024, i;          // For arbitrary function, n = 8192 for earthquake data
    double x[] = new double[n + 1];                                    // Data vector
    for ( i=1; i <= n; i++ )  {                // Initial data via arbitrary function
      x[i] = (1./1204)* (i*i-3.*i);  
      w.printf("%d\t%8.3f\n", i, x[i]); 
    }
     // deselect the following lines to process earthquake data
   /*  Scanner in = new Scanner(new File("earthquake.dat")); 
    i = 1; 
    while (in.hasNextDouble()) {a[i] = in.nextDouble(); 
			w.printf("%d\t%8.3f\n", i, a[i]); i++ ; 
    }     */
   pyram(x, n, 1);                                                             // DWT
   pyram(x, n, -1);                                                    // Inverse DWT
   for ( i=1; i <= n; i++ ) q.printf("%d\t%8.3f\n", i, x[i]); 
  }                                                                           // Main
}                                                                            // Class